llvm/lib/Target/NVPTX/NVPTX.h

```diff
@@ -106,15 +107,30 @@ enum LoadStore {
   isStoreShift = 6
 };
 
-namespace PTXLdStInstCode {
-enum MemorySemantic {
+// Extends LLVM AtomicOrdering with PTX Orderings:
+using OrderingUnderlyingType = unsigned int;
+enum Ordering : OrderingUnderlyingType {
   NotAtomic = 0, // PTX calls these: "Weak"
-  Volatile = 1,
+  // Unordered = 1, // NVPTX maps LLVM Unorderd to Relaxed
   Relaxed = 2,
-  Acquire = 3,
-  Release = 4,
-  RelaxedMMIO = 5
+  // Consume = 3,   // Unimplemented in LLVM; NVPTX would map to "Acquire"
+  Acquire = 4,
+  Release = 5,
+  // AcquireRelease = 6, // TODO
+  SequentiallyConsistent = 7,
+  Volatile = 8,
+  RelaxedMMIO = 9,
+  LAST = RelaxedMMIO
 };
+// Values match LLVM AtomicOrdering for common orderings:
+static_assert(Ordering::NotAtomic == (unsigned)AtomicOrdering::NotAtomic);
+static_assert(Ordering::Relaxed == (unsigned)AtomicOrdering::Monotonic);
+static_assert(Ordering::Acquire == (unsigned)AtomicOrdering::Acquire);
+static_assert(Ordering::Release == (unsigned)AtomicOrdering::Release);
+static_assert(Ordering::SequentiallyConsistent ==
+              (unsigned)AtomicOrdering::SequentiallyConsistent);
```

Comment:
This current version does the job, but it still rubs me the wrong way, a bit. We're explicitly verifying that the values match those of  AtomicOrdering, and *implicitly* relying on the numeric order of enum values by hardcoding them as literals.

I think it would make more sense to explicitly assign enum values to their AtomicOrdering counterparts, and then *explicitly* enforce the ordering via the static_assert. This would better express our intent here.

On the other hand, it's convenient to see the explicit values all in one place, and mixing AtomicOrdering constants and explicit literals would probably look a bit odd, too.

I'll leave it up to you.

Reply:
Not sure I fully understand what you are asking for yet. Is something like this what you have in mind:

```
enum Ordering : OrderingUnderlyingType {
  NotAtomic = AtomicOrdering::NotAtomic, // PTX calls these: "Weak"
  // Unordered = AtomicOrdering::Unordered, // NVPTX maps LLVM Unorderd to Relaxed
  Relaxed = AtomicOrdering::Monotonic,
  Acquire = AtomicOrdering::Acquire,
  Release = AtomicOrdering::Release,
  // AcquireRelease = AtomicOrdering::AcquireRelease,
  SequentiallyConsistent = AtomicOrdering::SequentiallyConsistent,
  Volatile = SequentiallyConsistent + 1,
  RelaxedMMIO = Volatile + 1,
  LAST = RelaxedMMIO
};
```

?

Reply:
Yeah, something like that. On a second thought, it's probably an overkill. While matching numeric values to AtomicOrdering is a convenience, it's not necessary. We're not extending AtomicOrdering, but, rather, defining an new ordering with some overlap with AtomicOrdering.

How important in practice is it to have the matching numeric values?

Perhaps what we need is a set of ordering values, w/o any attempt to inherit from AtomicOrdering, or comments that we map some values to something else, and, if needed, an explicit mapping function between AtomicOrdering and NVPTX ordering.

Reply:
I think its mostly a convenience. When debugging, I think its nice if 0 means the same thing both in LLVM and in the NVPTX backend. This also allows casting the LLVM one into the NVPTX one (or just using either as a "machine code").

The LLVM values are explicitly set and don't change often (I don't think they've ever changed), so I am not too worried about how to do this (static cast like right now, vs explicitly setting the values to the LLVM enum values). But I do think that not having to remember that 3 means something in AtomicOrdering and something different in NVPTX ordering is worth it.